File: frontend/src/types/route.ts
export interface Coordinate {
  lat: number;
  lon: number;
}

export interface RouteLeg {
  mode: string;
  route_short_name: string | null;
  headsign: string | null;
  from_name: string;
  to_name: string;
  from_coord: Coordinate;
  to_coord: Coordinate;
  distance_meters: number;
  duration_seconds: number;
  departure_time: string | null;
  arrival_time: string | null;
}

export interface RouteItinerary {
  duration_seconds: number;
  start_time: string;
  end_time: string;
  transfers: number;
  walk_distance_meters: number;
  legs: RouteLeg[];
}

export interface RouteRequest {
  from_lat: number;
  from_lon: number;
  to_lat: number;
  to_lon: number;
  departure_time?: string;
}

export interface RouteResponse {
  provider: string;
  itineraries: RouteItinerary[];
}

File: backend/app/schemas/route.py
from typing import Optional
from pydantic import BaseModel


class Coordinate(BaseModel):
    lat: float
    lon: float


class RouteLeg(BaseModel):
    mode: str           # WALK, BUS, TRAM, SUBWAY, RAIL
    route_short_name: Optional[str] = None
    headsign: Optional[str] = None
    from_name: str
    to_name: str
    from_coord: Coordinate
    to_coord: Coordinate
    distance_meters: float
    duration_seconds: int
    departure_time: Optional[str] = None  # ISO8601
    arrival_time: Optional[str] = None


class RouteItinerary(BaseModel):
    duration_seconds: int
    start_time: str
    end_time: str
    transfers: int
    walk_distance_meters: float
    legs: list[RouteLeg]


class RouteRequest(BaseModel):
    from_lat: float
    from_lon: float
    to_lat: float
    to_lon: float
    departure_time: Optional[str] = None   # ISO8601; None = now


class RouteResponse(BaseModel):
    provider: str
    itineraries: list[RouteItinerary]
